### DISCUSSION TITLE
Finding Values #2

### DISCUSSION THREAD

**Discussion by D8bp8Ags** [joalla/discogs_client#52] [Support] [answered]
Hi,

How can i access values that are not predefined in the class Release?

In the json result there are much more values then the Release class return. I read that not all values are mapped but how can i map them myself?
I added 'released' to the release class and that is working (not sure if this is the right way) but how can i get childeren from 'tracklist' which are present in the class Track?

some part returned from discogs json:

```
..
..
"released": "2020-02-07",
..
..
"tracklist": [
  {
    "position": "1",
    "type_": "track",
    "title": "DT64",
    "duration": "6:57"
  }
],
```
```
releases = self.ds.release(id)

print(releases.id)
print(releases.artists)
print(releases.released)
print(releases.duration) <<
```
models.py:
```
class Release(PrimaryAPIObject):
    id = SimpleField()
    title = SimpleField()
    year = SimpleField()
    thumb = SimpleField()
added: 
    released = SimpleField() << working

    <duration> = ListField('Track')....duration 
```
```
class Track(SecondaryAPIObject):
    duration = SimpleField()
    position = SimpleField()
    title = SimpleField()
    artists = ListField('Artist')
    credits = ListField('Artist', key='extraartists')

    def __repr__(self):
        return '<Track {0!r} {1!r}>'.format(self.position, self.title)
```

**Comment by JOJ0** [comment, marked as answer]
> Hi,
> 
> 
> 
> How can i access values that are not predefined in the class Release?
> 
> 
> 
> In the json result there are much more values then the Release class return. I read that not all values are mapped but how can i map them myself?
> 

They are already mapped in discogs_client, you just need to dig one object level deeper. The tracklist is a list you can loop through. The track items you'll find there have the properties you are looking for.

Have a look at our brand new documentation pages, PR is not through yet but they are accessible already. Here is exactly what you are looking for: https://python3-discogs-client.readthedocs.io/en/latest/fetching_data.html#release

As said, docs are pretty new. Feedback and pointing to badly explained things is more than welcome. Thanks!

**Comment by D8bp8Ags** [comment]
Hi,

Thanks for your reply.
```
Get details of the first track on this Release

release.tracklist[0].title
[...]
release.tracklist[0].duration
[...]
```
This is working.



> As said, docs are pretty new. Feedback and pointing to badly explained things is more than welcome. Thanks!

`Find the available properties of a Track object in the module docs: {class}discogs_client.models.Track`
Think this is a layout thing.

I compared different json results and there is no consistence.
My goal is to search on a release and generate a tracklist;

1 artist - song (orginal mix)
2 artist - song (radio edit)
3 artist - song (artist2 remix)


a few json examples with cuts:

example1:
```
{
  "id": 14759887,
  "status": "Accepted",
  "year": 2020,
  "resource_url": "https://api.discogs.com/releases/14759887",
  "uri": "https://www.discogs.com/Moguai-vs-Kai-Tracid-DT64/release/14759887",
  "artists": [
    {
      "name": "Moguai",
      "anv": "",
      "join": "vs",
      "role": "",
      "tracks": "",
      "id": 9439,
      "resource_url": "https://api.discogs.com/artists/9439"
    },
    {
      "name": "Kai Tracid",
      "anv": "",
      "join": "",
      "role": "",
      "tracks": "",
      "id": 11399,
      "resource_url": "https://api.discogs.com/artists/11399"
    }
  ],

  ..
  ..
  ..

  "tracklist": [
    {
      "position": "1",
      "type_": "track",
      "title": "DT64",
      "duration": "6:57"
    }
  ],
  "extraartists": [],
```
example2:
```
{
  "id": 16969989,
  "status": "Accepted",
  "year": 2020,
  "resource_url": "https://api.discogs.com/releases/16969989",
  "uri": "https://www.discogs.com/DamienJ-TechnoWorkxx/release/16969989",
  "artists": [
    {
      "name": "DamienJ",
      "anv": "",
      "join": "",
      "role": "",
      "tracks": "",
      "id": 7182837,
      "resource_url": "https://api.discogs.com/artists/7182837"
    }
  ],

  ..
  ..
  ..

  "tracklist": [
    {
      "position": "1",
      "type_": "track",
      "title": "TechnoWorkxx (Original Mix)",
      "duration": "7:22"
    },
    {
      "position": "2",
      "type_": "track",
      "title": "TechnoWorkxx",
      "extraartists": [
        {
          "name": "MaKaJa Gonzales",
          "anv": "",
          "join": "",
          "role": "Remix",
          "tracks": "",
          "id": 5553081,
          "resource_url": "https://api.discogs.com/artists/5553081"
        }
      ],
      "duration": "5:57"
    }
  ],
  "extraartists": [],
```
example3:
```
{
  "id": 15306505,
  "status": "Accepted",
  "year": 2020,
  "resource_url": "https://api.discogs.com/releases/15306505",
  "uri": "https://www.discogs.com/Twins-Project-Bass-In-Your-Face/release/15306505",
  "artists": [
    {
      "name": "Twins Project",
      "anv": "",
      "join": "",
      "role": "",
      "tracks": "",
      "id": 409755,
      "resource_url": "https://api.discogs.com/artists/409755"
    }
  ],

  ..
  ..
  ..

  "tracklist": [
    {
      "position": "1",
      "type_": "track",
      "title": "Bass In Your Face (Original Mix)",
      "duration": "6:24"
    },
    {
      "position": "2",
      "type_": "track",
      "title": "Bass In Your Face (Space 92 Remix)",
      "extraartists": [
        {
          "name": "Space 92",
          "anv": "",
          "join": "",
          "role": "Remix",
          "tracks": "",
          "id": 7881215,
          "resource_url": "https://api.discogs.com/artists/7881215"
        }
      ],
      "duration": "7:08"
    }
  ],
  "extraartists": [],      
```
example1:
loop on "artists": [ +  "tracklist": [ title
result:
```
1 - Moguai, Kai Tracid - DT64
```

example2:
loop on   "artists": [ + "tracklist": [ title
result:
```
1 - DamienJ - TechnoWorkxx (Original Mix)
2 - DamienJ - TechnoWorkxx (MaKaJa Gonzales Remix) ("extraartists" name + role)
```

example3:
loop on "artists": [ + "tracklist": [ title
result:
```
1 - Twins Project - Bass In Your Face (Original Mix)
2 - Twins Project - Bass In Your Face (Space 92 Remix) Space 92 Remix (+ "extraartists" name + role)
```

As you can see sometimes there is duplicate info in the track title example; (Bass In Your Face (Space 92 Remix) so you get duplicate information. Sometimes they use extraartists for a remix.

The json results look inconsistent. How can i generate a tracklist?

**Reply by JOJ0** [reply]
Hi, if thats's what the discogs api gives you, you'd have to deal with it in your code. discogs_client is just a frontend to it.

but i have a completely new approach for you: I have built a tool already for managing playlists: https://discodos.jojotodos.net - please give it a try and if features are missing you could send a PR and we improve it together.

**Reply by JOJ0** [reply]
also many thanks for pointing out the broken link to the track class. will fix tomorrow, for now this is the link with all the properties: https://python3-discogs-client.readthedocs.io/en/latest/discogs_client.html#discogs_client.models.Track

...update: link fixed. thanks for the hint!

**Comment by JOJ0** [comment]
I did a quick test - added the mentioned releases to my collection, created a new mix named "Album Name" and added some of the tracks on those releases to it. This is how DiscoDOS would render it:

```
(discodos) jojo@gin ~/git/discodos (master) $ disco mix "Album N" -v
  Mix  Name        Created              Updated              Played    Venue
    2  Album Name  2021-05-10 07:31:45  2021-05-10 08:14:24

|   # | Release           | Track               | Track                  |   Trk | Key   | BPM   | Key     | Trans.   | Trans.   | Track   |
|     |                   | Artist              | Name                   |    No |       |       | Notes   | Rating   | Notes    | Notes   |
|----:|:------------------|:--------------------|:-----------------------|------:|:------|:------|:--------|:---------|:---------|:--------|
|   1 | DT64              | Moguai / Kai Tracid | DT64                   |     1 |       |       |         |          |          |         |
|   2 | TechnoWorkxx      | DamienJ             | TechnoWorkxx (Original |     1 |       |       |         |          |          |         |
|   3 | Bass In Your Face | Twins Project       | Bass In Your Face      |     1 |       |       |         |          |          |         |
|     |                   |                     | (Original Mix)         |       |       |       |         |          |          |         |
|   4 | Bass In Your Face | Twins Project       | Bass In Your Face      |     2 |       |       |         |          |          |         |
|     |                   |                     | (Space 92 Remix)       |       |       |       |         |          |          |         |
```

I am not sure what your goal is exactly. You said you want to create playlists of Albums - DiscoDOS is actually ment for creating playlists out of single tunes on any release - DJ playlists/mixes. If what you are after is lists containing exactly what is on one release this surely can be workarounded with the current version of the tool by creating a mix and then adding all the tracks of one album/release to it. If that's what you'd like to do regularly then a feature needs to be added that quickly adds whole releases to a mix/playlist. 

I sent you a PM on discogs.com - Since this gets out of scope of your question here, let's continue chatting as suggested in my PM.

Also I'l mark your question here as anwered since your issue with finding details of tracks in tracklists is solved. You agree?

**Reply by D8bp8Ags** [reply]
Thanks for your reply and info. It's okay to close the issue
